test/prefer-flat-map.js

```diff
@@ -0,0 +1,105 @@
+import test from 'ava';
+import avaRuleTester from 'eslint-ava-rule-tester';
+import rule from '../rules/prefer-flat-map';
+
+const ruleTester = avaRuleTester(test, {
+	env: {
+		es6: true
+	}
+});
+
+const error = {
+	ruleId: 'prefer-flat-map',
+	messageId: 'preferFlatMap'
+};
+
+ruleTester.run('prefer-flat-map', rule, {
+	valid: [
+		'const bar = [1,2,3].map()',
+		'const bar = [1,2,3].map(i => i)',
+		'const bar = [1,2,3].map((i) => i)',
+		'const bar = [1,2,3].map((i) => { return i; })',
+		'const bar = foo.map(i => i)',
+		'const bar = [[1],[2],[3]].flat()',
+		'const bar = [1,2,3].map(i => [i]).sort().flat()',
+		`
+			let bar = [1,2,3].map(i => [i]);
+			bar = bar.flat();
+		`
+	],
+	invalid: [
+		{
+			code: 'const bar = [1,2,3].map(i => [i]).flat()',
+			output: 'const bar = [1,2,3].flatMap(i => [i])',
+			errors: [error]
+		},
+		{
+			code: 'const bar = [1,2,3].map((i) => [i]).flat()',
+			output: 'const bar = [1,2,3].flatMap((i) => [i])',
+			errors: [error]
+		},
+		{
+			code: 'const bar = [1,2,3].map((i) => { return [i]; }).flat()',
+			output: 'const bar = [1,2,3].flatMap((i) => { return [i]; })',
+			errors: [error]
+		},
+		{
+			code: 'const bar = [1,2,3].map(foo).flat()',
+			output: 'const bar = [1,2,3].flatMap(foo)',
+			errors: [error]
+		},
+		{
+			code: 'const bar = foo.map(i => [i]).flat()',
+			output: 'const bar = foo.flatMap(i => [i])',
+			errors: [error]
+		},
+		{
+			code: 'const bar = { map: () => {} }.map(i => [i]).flat()',
+			output: 'const bar = { map: () => {} }.flatMap(i => [i])',
+			errors: [error]
+		},
+		{
+			code: 'const bar = [1,2,3].map(i => i).map(i => [i]).flat()',
+			output: 'const bar = [1,2,3].map(i => i).flatMap(i => [i])',
+			errors: [error]
+		},
+		{
+			code: 'const bar = [1,2,3].sort().map(i => [i]).flat()',
+			output: 'const bar = [1,2,3].sort().flatMap(i => [i])',
+			errors: [error]
+		},
+		{
+			code: 'const bar = (([1,2,3].map(i => [i]))).flat()',
+			output: 'const bar = (([1,2,3].flatMap(i => [i])))',
```

Comment:
I see that the fix range logic is really tricky, can you also add tests for different spacing patterns like these?

```js
array
	.map(...)
	.flat()
```

```js
array.map(...)
	.flat()
```

```js
array . map( x => y ) . flat () // 🤪
```